bpfd-operator/controllers/bpfd-agent/uprobe-program_test.go

```diff
@@ -0,0 +1,211 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+func TestUprobeProgramControllerCreate(t *testing.T) {
+	var (
+		name            = "fakeUprobeProgram"
+		namespace       = "bpfd"
+		bytecodePath    = "/tmp/hello.o"
+		sectionName     = "test"
+		functionName    = "malloc"
+		target          = "libc"
+		offset          = 0
+		retprobe        = false
+		uprobenamespace = ""
+		fakeNode        = testutils.NewNode("fake-control-plane")
+		ctx             = context.TODO()
+		bpfProgName     = fmt.Sprintf("%s-%s-%s", name, fakeNode.Name, "libc")
+		bpfProg         = &bpfdiov1alpha1.BpfProgram{}
+		fakeUID         = "ef71d42c-aa21-48e8-a697-82391d801a81"
+	)
+	// A UprobeProgram object with metadata and spec.
+	Uprobe := &bpfdiov1alpha1.UprobeProgram{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: bpfdiov1alpha1.UprobeProgramSpec{
+			BpfProgramCommon: bpfdiov1alpha1.BpfProgramCommon{
+				SectionName:  sectionName,
+				NodeSelector: metav1.LabelSelector{},
+				ByteCode: bpfdiov1alpha1.BytecodeSelector{
+					Path: &bytecodePath,
+				},
+			},
+			FunctionName: functionName,
+			Targets:      []string{target},
+			Offset:       uint64(offset),
+			RetProbe:     retprobe,
+		},
+	}
+
+	// Objects to track in the fake client.
+	objs := []runtime.Object{fakeNode, Uprobe}
+
+	// Register operator types with the runtime scheme.
+	s := scheme.Scheme
+	s.AddKnownTypes(bpfdiov1alpha1.SchemeGroupVersion, Uprobe)
+	s.AddKnownTypes(bpfdiov1alpha1.SchemeGroupVersion, &bpfdiov1alpha1.UprobeProgramList{})
+	s.AddKnownTypes(bpfdiov1alpha1.SchemeGroupVersion, &bpfdiov1alpha1.BpfProgram{})
+	s.AddKnownTypes(bpfdiov1alpha1.SchemeGroupVersion, &bpfdiov1alpha1.BpfProgramList{})
+
+	// Create a fake client to mock API calls.
+	cl := fake.NewClientBuilder().WithRuntimeObjects(objs...).Build()
+
+	cli := agenttestutils.NewBpfdClientFake()
+
+	rc := ReconcilerCommon{
+		Client:       cl,
+		Scheme:       s,
+		BpfdClient:   cli,
+		NodeName:     fakeNode.Name,
+		expectedMaps: map[string]string{},
+	}
+
+	// Set development Logger so we can see all logs in tests.
+	logf.SetLogger(zap.New(zap.UseFlagOptions(&zap.Options{Development: true})))
+
+	// Create a ReconcileMemcached object with the scheme and fake client.
+	r := &UprobeProgramReconciler{ReconcilerCommon: rc, ourNode: fakeNode}
+
+	// Mock request to simulate Reconcile() being called on an event for a
+	// watched resource .
+	req := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+
+	// First reconcile should create the bpf program object
+	res, err := r.Reconcile(ctx, req)
+	if err != nil {
+		t.Fatalf("reconcile: (%v)", err)
+	}
+
+	// Check the BpfProgram Object was created successfully
```

Comment:
Additionally for all this testing infra... I wonder if we can condense it somehow.... (followup)

Reply:
I created https://github.com/bpfd-dev/bpfd/issues/710